### PR TITLE
GrafanaUI: Expose Text component

### PR DESCRIPTION
**Pull request by joshhunt**
The main piece of feedback [I have after using the new text components](https://github.com/grafana/grafana/pull/68532) is that `<TextVariant />` has an unweidly name, it doesn't do enough, and adding the rest of the props would just make it the same as Text anyway.

 - Exposes Text component
   - Text component can only be used with `body` and `bodySmall` variants. To get a heading variant, using a Heading component
 - Adds fontStyle for italic
     - 🤔 Wasn't sure about exposing `fontStyle`, or having a dedicated `italic` prop. Thoughts?
 - Uses 'function factory' to create Heading elements
 - Removed Span, Legend, and TextVariant components in favor of just using Text component directly
    - After exposing Text, Span and TextVariant become redundant. 
    - Span isn't really a semantic element - it's mainly just used when we want to style some inline text... which is what Text/TextVariant is 😅 
     - I did go a bit cowboy with removing Legend - it _felt_ like it was made just to expose `bodySmall` variant? Which can just be used with Text now. I think its much easier to start with less and add more later, than to have a component hanging around that we dont use.
 - Moves getStyles into function to reuse between typography components 

Usage:

```ts

<H2 variant="h1">Welcome back <Text weight="ultraBold">Josh</Text><H2>

<Text color="secondary" fontStyle="italic>No items</Text>

```

(ignore that we don't actually have an ultra-bold weight)

---

It's becoming more certain that I dislike the concept of an `as` prop:

 - `as` is a common convention for component libraries, but I think `<Text as="h3">Hello<Text/>` is confusing - does `as` control the HTML element, the variant appearance, or both?
 - I think props should do less things and be less magic - explicitness over convention.
 - Renaming it to something like `element` or `component` fixes this, but _to me_ it seems weird for a high-level component to have a `element` prop when we already have a way to communicate that information - the tag name!
 - A similar use case is Button vs LinkButton - we don't give them an `as` prop to control the element and instead we make a pre-packaged element
 - If we say that the behaviour of `as` is learnable (which is fair) if not untuitive, then surely we can say that knowing to use `<H1/>`, `<H2/>`, `<H3/>`, and `<Text />` is equally learnable, with the advantage of "the component tells you the element"?

I think `as` (or a different name) is fine for like an internal component that people can 'drop down to' infrequently if something higher-level doesn't fulfil their use case (`Box` component is a good example of this I think), but should not be a part of the common 'happy path' API for compoennts.

### CODE REVIEW

File: packages/grafana-ui/src/components/Text/Text.tsx [hunk 1]
```diff
@@ -1,106 +1,37 @@
...
-  /** Color to use for text */
-  color?: keyof GrafanaTheme2['colors']['text'] | 'error' | 'success' | 'warning' | 'info';
-  /** Use to cut the text off with ellipsis if there isn't space to show all of it. On hover shows the rest of the text */
-  truncate?: boolean;
-  /** Whether to align the text to left, center or right */
-  textAlignment?: CSSProperties['textAlign'];
+export interface TextProps extends TextStyleProps {
   children: React.ReactNode;
+
+  /** What typograpy variant should be used for the component. Only use if default variant for the defined element is not what is needed */
```
Comment (joshhunt) [inline packages/grafana-ui/src/components/Text/Text.tsx:10]: comment needs updating

**Review by JoaoSilvaGrafana — COMMENTED** [review]
I'm happy with pretty much all of this including dropping `Legend`, it was a confusing name anyway.

One comment I left here though is about what does `Text` mean now? If we have "element" components and then this less specific component I think we are mixing two different concepts. If I looked at this list of components: `H1`, `H2`, `H3`, `H4`, `H5`, `H6`, `P` and `Text` I would be confused at what `Text` does because its the odd one out.

As I left in the comments I would then prefer to just go `Heading` and `Text` instead of this mix.

File: packages/grafana-ui/src/components/Text/styles.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+export interface TextStyleProps {
+  /** Override the default weight for the used variant */
+  weight?: 'light' | 'regular' | 'medium' | 'bold';
+
+  /** Color to use for text */
+  color?: keyof GrafanaTheme2['colors']['text'] | 'error' | 'success' | 'warning' | 'info';
+
+  /** Font style, such as italics  */
+  fontStyle?: CSSProperties['fontStyle'];
```
[the remark this answers is not in the document]
Comment (joshhunt) [reply packages/grafana-ui/src/components/Text/styles.ts:14]: My thinking exactly! What about just `italic`?

```ts
<Text color="secondary" fontWeight="bold" italic>No items</Text>
```

File: packages/grafana-ui/src/components/Text/Text.tsx [hunk 3]
```diff
@@ -1,106 +1,37 @@
...
+  variant?: 'body' | 'bodySmall';
 }
 
 export const Text = React.forwardRef<HTMLElement, TextProps>(
```
[the remark this answers is not in the document]
Comment (JoaoSilvaGrafana) [reply packages/grafana-ui/src/components/Text/Text.tsx:14]: Sorry didn't realize there was a reply!

I guess my first thought when I saw a component called `Text` was the original component we created that basically lets you manipulate it as you wanted, while this is more the concept of a `Text` as opposed to `Heading`. I think there will be a bit of learning to do at the start to understand how to use it. But I agree that if we dropped `P` that might solve some of this confusion. Also maybe not letting any of the headings be styled with `body` variants? Then there is a more clear separation of the two ideas?

**Pull request closed without merging**